server/queue/queue-rabbitmq/src/main/java/org/apache/james/queue/rabbitmq/managable/cassandra/EnqueuedMailsDao.java

```diff
@@ -0,0 +1,391 @@
+
+        this.firstEnqueuedMailDao = firstEnqueuedMailDao;
+        this.deletedMailsDao = deletedMailsDao;
+        this.executor = new CassandraAsyncExecutor(session);
+        this.cassandraUtils = cassandraUtils;
+        this.cassandraTypesProvider = cassandraTypesProvider;
+        this.manageableQueueConfiguration = manageableQueueConfiguration;
+
+        this.selectFrom = prepareSelectFrom(session);
+        this.insert = prepareInsert(session);
+    }
+
+    private PreparedStatement prepareSelectFrom(Session session) {
+        return session.prepare(select()
+                .from(TABLE_NAME)
+                .where(eq(QUEUE_NAME, bindMarker(QUEUE_NAME)))
+                .and(gte(TIME_RANGE_START, bindMarker(TIME_RANGE_START)))
+                .and(eq(BUCKET_ID, bindMarker(BUCKET_ID))));
+    }
+
+    private PreparedStatement prepareInsert(Session session) {
+        return session.prepare(insertInto(TABLE_NAME)
+            .value(MAIL_KEY, bindMarker(MAIL_KEY))
+            .value(STATE, bindMarker(STATE))
+            .value(SENDER, bindMarker(SENDER))
+            .value(RECIPIENTS, bindMarker(RECIPIENTS))
+            .value(ATTRIBUTES, bindMarker(ATTRIBUTES))
+            .value(ERROR_MESSAGE, bindMarker(ERROR_MESSAGE))
+            .value(REMOTE_ADDR, bindMarker(REMOTE_ADDR))
+            .value(REMOTE_HOST, bindMarker(REMOTE_HOST))
+            .value(LAST_UPDATED, bindMarker(LAST_UPDATED))
+            .value(PER_RECIPIENT_SPECIFIC_HEADERS, bindMarker(PER_RECIPIENT_SPECIFIC_HEADERS)));
+    }
+
+    public CompletableFuture<Void> enqueuedMail(Mail mail, MailQueueName mailQueueName) {
+        return insert(mail, mailQueueName)
+            .thenCompose(aVoid -> firstEnqueuedMailDao.updateEnqueuedMail(mailQueueName));
+    }
+
+    private CompletableFuture<Void> insert(Mail mail, MailQueueName mailQueueName) {
+        return executor.executeVoid(insert.bind()
+            .setString(QUEUE_NAME, mailQueueName.asString())
+            .setTimestamp(TIME_RANGE_START, Date.from(currentSliceStartInstant()))
+            .setInt(BUCKET_ID, randomBucketId(mail))
+            .setString(MAIL_KEY, mail.getName())
+            .setString(STATE, mail.getState())
+            .setString(SENDER, Optional.ofNullable(mail.getSender())
+                .map(MailAddress::asString)
+                .orElse(null))
+            .setList(RECIPIENTS, asStringList(mail.getRecipients()))
+            .setString(ERROR_MESSAGE, mail.getErrorMessage())
+            .setString(REMOTE_ADDR, mail.getRemoteAddr())
+            .setString(REMOTE_HOST, mail.getRemoteHost())
+            .setTimestamp(LAST_UPDATED, mail.getLastUpdated())
+            .setMap(ATTRIBUTES, toRawAttributeMap(mail))
+            .setMap(PER_RECIPIENT_SPECIFIC_HEADERS, toHeaderMap(cassandraTypesProvider, mail.getPerRecipientSpecificHeaders()))
+        );
+    }
+
+    public CompletableFuture<Stream<ManageableMailQueue.MailQueueItemView>> browse(
```

Comment:
IMO to be a real DAO the caller should pass all the BucketedSlices he wants to read.